pymeasure/instruments/teledyne/teledyne_oscilloscope.py

```diff
@@ -0,0 +1,1136 @@
+        The WRITE_INTERVAL_S is set to 10ms as default however its optimal value heavily depends
+        on the actual commands and on the connection type, so it is impossible to give a unique
+        value to fit all cases. An interval between 10ms and 500ms second proved to be good,
+        depending on the commands and connection latency.
+    """
+
+    _BOOLS = TeledyneOscilloscopeChannel._BOOLS
+
+    WRITE_INTERVAL_S = 0.02  # seconds
+
+    ch_1 = Instrument.ChannelCreator(TeledyneOscilloscopeChannel, 1)
+
+    ch_2 = Instrument.ChannelCreator(TeledyneOscilloscopeChannel, 2)
+
+    ch_3 = Instrument.ChannelCreator(TeledyneOscilloscopeChannel, 3)
+
+    ch_4 = Instrument.ChannelCreator(TeledyneOscilloscopeChannel, 4)
+
+    def __init__(self, adapter, name="Teledyne Oscilloscope", **kwargs):
+        super().__init__(adapter, name=name, **kwargs)
+        if self.adapter.connection is not None:
+            self.adapter.connection.timeout = 3000
+        self._grid_number = 14  # Number of grids in the horizontal direction
+        self._seconds_since_last_write = 0  # Timestamp of the last command
+        self._header_size = 16  # bytes
+        self._footer_size = 2  # bytes
+        self.waveform_source = "C1"
+        self.default_setup()
+
+    ################
+    # System Setup #
+    ################
+
+    def default_setup(self):
+        """ Set up the oscilloscope for remote operation.
+
+        The COMM_HEADER command controls the
+        way the oscilloscope formats response to queries. This command does not affect the
+        interpretation of messages sent to the oscilloscope. Headers can be sent in their long or
+        short form regardless of the CHDR setting.
+        By setting the COMM_HEADER to OFF, the instrument is going to reply with minimal
+        information, and this makes the response message much easier to parse.
+        The user should not be fiddling with the COMM_HEADER during operation, because
+        if the communication header is anything other than OFF, the whole driver breaks down.
+        """
+        self._comm_header = "OFF"
+
+    def ch(self, source):
+        """ Get channel object from its index or its name. Or if source is "math", just return the
+        scope object.
+
+        :param source: can be 1, 2, 3, 4 or C1, C2, C3, C4, MATH
+        :return: handle to the selected source.
+        """
+        if isinstance(source, str):
+            source = sanitize_source(source)
+        if source == "MATH":
+            return self
+        elif source == "LINE":
+            raise ValueError("LINE is not a valid channel")
+        else:
+            return getattr(self, f"ch_{source if isinstance(source, int) else source[-1]}")
+
+    def autoscale(self):
+        """ Autoscale displayed channels."""
+        self.write("ASET")
+
+    def write(self, command, **kwargs):
+        """ Writes the command to the instrument through the adapter.
```

Comment:
Just a small thing, as we are touching the whole file:
```suggestion
        """ Write the command to the instrument through the adapter.
```
(docstrings should have their first phrase in imperative mode.)